### PR TITLE
Restrict user to drop a column that is used in older partition spec or void partition column in iceberg connector

### PR DESCRIPTION
**Pull request by krvikash**
## Description
Restrict the user to drop a column that is used in older partition spec or void partition column in iceberg connector until the below issues got fixed.

This PR changes the older behavior for the void partition column. As earlier users could drop a void partition column. But now it is not permissible. CC: @ebyhr 

https://github.com/trinodb/trino/issues/15729
https://github.com/trinodb/trino/issues/15738

## Release notes

(X) This is not user-visible or docs only and no release notes are required.

### CODE REVIEW

**Comment by ebyhr** [comment]
Can we add a test?

**Comment by findepi** [comment]
> until the below issues got fixed.

would it be hard to fix https://github.com/trinodb/trino/issues/15729 https://github.com/trinodb/trino/issues/15738 ?

**Comment by krvikash** [comment]
https://github.com/trinodb/trino/issues/15738 is partially fixed.

Only https://github.com/trinodb/trino/issues/15729 needs to be fixed. The issue is from iceberg library. I have to check the feasibility.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 1]
```diff
@@ -1532,10 +1531,20 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
...
         }
+        int currentSpecId = icebergTable.spec().specId();
+        boolean columnUsedInOlderPartitionSpecs = icebergTable.specs().entrySet().stream()
+                .filter(spec -> spec.getValue().specId() != currentSpecId)
+                .flatMap(spec -> spec.getValue().fields().stream())
+                .anyMatch(field -> field.sourceId() == handle.getId());
+        if (columnUsedInOlderPartitionSpecs) {
+            // After dropping a column which was used in older partition specs, insert/update/select fails on the table.
+            // So restricting user to dropping that column. https://github.com/trinodb/trino/issues/15729
+            throw new TrinoException(NOT_SUPPORTED, "Cannot drop column '" + handle.getName() + "' that is used in older partition spec");
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1546]: Grammar nitpick, and usually it's preferred to have variable names at the end of the error message:
```suggestion
            throw new TrinoException(NOT_SUPPORTED, "Cannot drop column which is used by an old partition spec: " + handle.getName());
```

**Review by alexjo2144 — APPROVED** [review]
Thanks!

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -1544,7 +1543,7 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
         if (columnUsedInOlderPartitionSpecs) {
             // After dropping a column which was used in older partition specs, insert/update/select fails on the table.
             // So restricting user to dropping that column. https://github.com/trinodb/trino/issues/15729
-            throw new TrinoException(NOT_SUPPORTED, "Cannot drop column '" + handle.getName() + "' that is used in older partition spec");
+            throw new TrinoException(NOT_SUPPORTED, "Cannot drop column which is used by an old partition spec: " + handle.getName());
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:1546]: This change should go to the 1st commit.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 3]
```diff
@@ -1225,9 +1225,9 @@ public void testDropColumnUsedInOlderPartitionSpecs()
         String tableName = "test_drop_partition_column_" + randomNameSuffix();
         assertUpdate("CREATE TABLE " + tableName + " (id INTEGER, name VARCHAR, age INTEGER) WITH (partitioning = ARRAY['id', 'truncate(name, 5)', 'void(age)'])");
         assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES partitioning = ARRAY[]");
-        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN id", "Cannot drop column 'id' that is used in older partition spec");
-        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN name", "Cannot drop column 'name' that is used in older partition spec");
-        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN age", "Cannot drop column 'age' that is used in older partition spec");
+        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN id", "Cannot drop column which is used by an old partition spec: id");
+        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN name", "Cannot drop column which is used by an old partition spec: name");
+        assertQueryFails("ALTER TABLE " + tableName + " DROP COLUMN age", "Cannot drop column which is used by an old partition spec: age");
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1230]: Move to 1st commit.

**Pull request merged**